scipy/stats/_hypotests.py

```diff
@@ -663,3 +665,217 @@ def somersd(x, y=None):
+    With `barnard_exact`, the probability that we would observe this or an
+    even more imbalanced ratio by chance is about 3.4%. With the same
+    significance level as above, we can therefore conclude that our
+    observed imbalance is statistically significant; The vaccine has a
+    positive effect.
+    """
+    if num_it < 0:
+        raise ValueError(
+            "Number of iterations `num_it` must be non-negative, "
+            f"found {num_it!r}")
+
+    c = np.asarray(
+        table,
+        dtype=np.int64)
+
+    if not c.shape == (2, 2):
+        raise ValueError("The input `table` must be of shape (2, 2).")
+
+    if np.any(c < 0):
+        raise ValueError("All values in `table` must be nonnegative.")
+
+    if 0 in c.sum(axis=0):
+        # If both values in column are zero, the p-value is 1 and
+        # the score's statistic is NaN.
+        return BarnardExactResult(np.nan, 1.0)
+
+    total_c1, total_c2 = c.sum(axis=0)
+    n = total_c1 + total_c2
+
+    x1 = np.arange(total_c1 + 1, dtype=np.int64).reshape(-1, 1)
+    x2 = np.arange(total_c2 + 1, dtype=np.int64).reshape(1, -1)
+
+    p1, p2 = x1 / total_c1, x2 / total_c2
+
+    if pooled:
+        p = (x1 + x2) / (total_c1 + total_c2)
+        var_p1_p2 = p * (1 - p) * (1 / total_c1 + 1 / total_c2)
+    else:
+        var_p1_p2 = p1 * (1 - p1) / total_c1 + p2 * (1 - p2) / total_c2
+
+    # To avoid warning when dividing by 0
+    with np.errstate(divide='ignore', invalid='ignore'):
+        TX = np.divide((p2 - p1), np.sqrt(var_p1_p2))
+
+    TX[p1 == p2] = 0  # Removing NaN values
+
+    TX_obs = TX[c[0, 0], c[0, 1]]
+
+    if alternative == "two-sided":
+        idx = np.abs(TX) >= abs(TX_obs)
+    elif alternative == "less":
+        idx = TX <= - abs(TX_obs)
+    elif alternative == "greater":
+        idx = TX >= abs(TX_obs)
+    else:
+        msg = (
+            "`alternative` should be one of {'two-sided', 'less', 'greater'},"
+            f" found {alternative!r}")
+        raise ValueError(msg)
+
+    # Pass x1 and x2 in dimension 3
+    x1 = x1.reshape(-1, 1, 1)
+    x2 = x2.reshape(1, -1, 1)
+
+    x1_comb = _barnard_exact_compute_gammaln_combination(total_c1)
+    x2_comb = _barnard_exact_compute_gammaln_combination(total_c2)
+
+    nuisance_num = 100
+    inf_bound, sup_bound = 0, 1
+
+    for _ in range(num_it):
+        nuisance_arr = np.linspace(
+            start=inf_bound, stop=sup_bound, num=nuisance_num)
+        nuisance_arr = nuisance_arr
+        # Reshape in dimension 3 array
+        nuisance_arr = nuisance_arr.reshape(1, 1, -1)
+
+        with np.errstate(divide='ignore', invalid='ignore'):
+            tmp_log_nuisance = np.log(
+                        nuisance_arr, out=np.zeros_like(nuisance_arr),
+                        where=nuisance_arr >= 0)
+            tmp_log_1_minus_nuisance = (
+                np.log(1 - nuisance_arr, out=np.zeros_like(nuisance_arr),
+                        where=1 - nuisance_arr >= 0)
```

Comment:
```suggestion
                       where=1 - nuisance_arr >= 0)
```
Here and elsewhere align with the parentheses, please.